### PR TITLE
[#40, #54] Configuration file with computed fields and include/exclude file options

### LINKED ISSUE #40: Config file with basic content include/exclude options
## Acceptance **✅19-12-2023** #97 

A config file that allows for setting:
- [x] path to content directory (defaults to `./content`)
- [x] files/directories to include
- [x] files/directories to exclude
- [x] 🏆 `npx markdowndb init` CLI script that will bootstrap the config file ( This is not needed )

### LINKED ISSUE #54: Computed metadata fields
Computed fields (or just any operations on incoming records)

cf https://www.contentlayer.dev/docs/reference/source-files/define-document-type#computedfields

When loading a file i want to create new fields based on some computation so that i can have additional metadata

- Add a type based on the folder of the file so that i can label blog posts
- Add `layout` based on the folder so i can change layouts based on folder
- compute title frontmatter from first heading 

More Examples:

- Generate keywords based on the content of the file.
- Label blog posts based on the folder structure.
- Count the number of words in the content.
- Estimate the reading time based on the content.
- Detect the language of the content.

## Acceptance

- [x] Design sketch with API/UX for users of lib in README or similar
- [x] Simple example in examples folder just a single js file like `index.js`

```
import indexFolder from markdowndb

files = indexFolder(../../__mock__, computedFields=...)

console.log(files[0].readingTime)
```

## Tasks

 - [x] sketch out the design before implementing
 - [x] ...

## Design

We can achieve this by allowing users to define custom fields using JavaScript functions. Each function takes the following information as parameters:

- File path
- File metadata (e.g., `tags`, `title`, ...)
- File body content
- File type (e.g., `blog`, ...)

The custom field generated by the function can then be added to the document scheme.

### Example Implementation

```javascript
// User-defined function to generate keywords based on content
function generateKeywords(filePath, metadata, content, fileType) {
  const keywords = ....
  return { keywords };
}

// Document scheme with custom field added
const documentScheme = {
  keywords: { value: generateKeywords}, // this will be excuted
};
```

In this example, the `generateKeywords` function is a placeholder for the user-defined function. Users can implement similar functions to automate the generation of custom fields based on their specific requirements. The custom field is then added to the document scheme for validation and type decleration.

## Notes

### PR DESCRIPTION
**Pull request by mohamedsalem401**
Closes #40
Closes #54

- Integrate a configuration file options.
- Implement computed fields.
- Provide the ability to configure include and exclude file patterns similar to the approach used in Next.js/TypeScript.

### CODE REVIEW

**Comment by rufuspollock** [comment]
@mohamedsalem401 does this reflect the comments in the PR yesterday re the definition in the README PR of this feature?

File: src/bin/index.js [hunk 1]
```diff
@@ -1,21 +1,40 @@
 #!/usr/bin/env node
-
+import * as path from 'path';
+import { promises as fs } from 'fs';
```
Comment (rufuspollock) [inline src/bin/index.js:3]: fs is defined but not used.

File: src/bin/index.js [hunk 2]
```diff
@@ -1,21 +1,40 @@
...
+  } catch (error) {
+    throw new Error(`Error loading configuration file from ${normalizedPath}: ${error.message}`);
+  }
+}
+
 // TODO get these from markdowndb.config.js or something
 const dbPath = "markdown.db";
 const ignorePatterns = [/Excalidraw/, /\.obsidian/, /DS_Store/];
-const [contentPath] = process.argv.slice(2);
+const [contentPath, configFilePath] = process.argv.slice(2);
```
Comment (rufuspollock) [inline src/bin/index.js:22]: What happens if you don't have a config file - this stuff should work with defaults ...

**Review by rufuspollock — CHANGES_REQUESTED** [review]
I can already see at least one issue is that the system should continue to work *without* a config folder.

In general, i'd strongly recommend (as discussed last week) that we get an example working without the config folder stuff but just with passing config to the relevant class or function e.g. `buildMarkdownDB` or whatever it is called.

once that API is well shaped adding config in a file is relatively straightforward ...

**Comment by rufuspollock** [comment]
@mohamedsalem401 another general comment - i'd have kept the include/exclude feature for a separate PR. It makes the review easier and faster ...

**Comment by mohamedsalem401** [comment]
I will close this for now

**Pull request closed without merging**